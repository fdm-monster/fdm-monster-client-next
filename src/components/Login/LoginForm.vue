<template>
  <v-card
    class="pa-4"
    elevation="10"
    style="border-radius: 10px">
    <v-card-text>
      <v-form>
        <label>
          Username
        </label>
        <v-text-field
          v-model="username"
          autofocus
          name="login"
          placeholder="Username"
          prepend-icon="mdi-account"
          type="text"
          variant="underlined"
          @keyup.enter="formIsDisabled || login()" />
        <v-text-field
          id="password"
          v-model="password"
          :append-icon="showPassword ? 'visibility_off' : 'visibility'"
          :type="showPassword ? 'text' : 'password'"
          label="Password"
          name="password"
          password
          prepend-icon="lock"
          variant="underlined"
          @click:append="showPassword = !showPassword"
          @keyup.enter="formIsDisabled || login()" />
        <v-alert
          v-if="errorMessage"
          class="mt-6"
          color="error"
          density="compact"
          variant="outlined">
          {{ errorMessage }}
        </v-alert>
        <v-alert
          v-if="authStore.lastLogoutReason"
          class="mt-6"
          color="error-darken-1"
          density="compact"
          variant="outlined">
          Reason for automatic logout: {{ authStore.lastLogoutReason }}
        </v-alert>
      </v-form>
    </v-card-text>
    <v-card-actions>
      <v-btn
        :loading="loading"
        class="pa-4"
        color="primary"
        size="lg"
        variant="flat"
        style="width: 100%"
        @click="login()">
        Login
      </v-btn>
    </v-card-actions>
    <v-card-actions>
      <v-btn
        :disabled="!authStore.registration"
        class="pa-4"
        size="lg"
        variant="flat"
        style="width: 100%"
        @click="gotoRegistration()">
        Register new account {{ authStore.registration ? "" : "(not enabled)" }}
        <v-icon
          class="pl-5"
          icon="right" />
      </v-btn>
    </v-card-actions>
  </v-card>
</template>

<script lang="ts" setup>
import {computed, onMounted, ref} from 'vue'
import {useRoute, useRouter} from 'vue-router'
import {useEventBus} from '@vueuse/core'
import {AxiosError} from 'axios'
import {useAuthStore} from '@/store/auth.store'
import {useSnackbar} from '@/shared/snackbar.composable'
import {RouteNames} from '@/router/route-names'
import {AUTH_ERROR_REASON, convertAuthErrorReason} from '@/shared/auth.constants'

const authStore = useAuthStore()
const errorMessage = ref('')
const username = ref('')
const showPassword = ref(false)
const password = ref('')
const router = useRouter()
const route = useRoute()
const loading = ref(false)
const loginEvent = useEventBus('auth:login')
const snackbar = useSnackbar()

const formIsDisabled = computed(() => {
  return (username.value ?? '')?.length < 3 || (password.value ?? '').length < 3
})

onMounted(async () => {
  authStore.loadTokens()
  await authStore.checkAuthenticationRequirements()
  if (authStore.loginRequired === false) {
    // As AppLoader might not trigger, we trigger it ourselves
    console.debug('LoginView, no login required, redirecting to', route.query.redirect, 'or home')
    loginEvent.emit(true)
    return await routeToRedirect()
  }
  if (!authStore.hasRefreshToken) {
    return
  }

  // Check if login is already valid, if so route away safely
  const success = await authStore.verifyOrRefreshLoginOnceOrLogout()
  if (success) {
    await routeToRedirect()
  }
})

async function gotoRegistration() {
  return await router.push({name: RouteNames.Registration})
}

async function login() {
  try {
    loading.value = true
    await authStore.login(username.value, password.value)
    authStore.lastLogoutReason = null
    password.value = ''
    loading.value = false
  } catch (e) {
    loading.value = false
    if ((e as AxiosError)?.response?.status === 401) {
      password.value = ''

      const reasonCode: keyof typeof AUTH_ERROR_REASON = ((e as AxiosError)?.response?.data as any)
        ?.reasonCode
      const convertedReason = convertAuthErrorReason(reasonCode)
      if (reasonCode === AUTH_ERROR_REASON.AccountNotVerified) {
        snackbar.error(
          convertedReason,
          'Please ask your administrator to verify your account and try again.'
        )
      } else if (reasonCode === AUTH_ERROR_REASON.PasswordChangeRequired) {
        snackbar.error(
          convertedReason,
          'Your password needs to be changed. This feature is sadly not finished'
        )
      }

      errorMessage.value = convertedReason
      return
    }

    snackbar.openErrorMessage({
      title: 'Error logging in',
      subtitle: 'Please test your connection and try again.',
    })
    errorMessage.value = 'Error logging in - status code ' + (e as AxiosError)?.response?.status
    password.value = ''

    return
  }

  errorMessage.value = ''

  // Trigger AppLoader
  loginEvent.emit(true)

  return await routeToRedirect()
}

async function routeToRedirect() {
  const routePath = route.query.redirect
  if (!routePath) {
    console.debug('[LoginForm] Redirecting to home')
    await router.push({name: RouteNames.Home})
    return
  } else {
    console.debug('[LoginForm] Redirecting to ', routePath)
    await router.push({
      path: routePath as string,
    })
    return
  }
}
</script>
