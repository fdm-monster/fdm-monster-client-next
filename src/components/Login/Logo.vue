<template>
  <v-img
    :src="imgUrl"
    class="shrink mr-1 pt-3 ml-1"
    contain
    style="opacity: 0.85"
    width="150"
  ></v-img>
</template>

<script lang="ts" setup>
import imgUrl from '@/assets/logo.png'
</script>
