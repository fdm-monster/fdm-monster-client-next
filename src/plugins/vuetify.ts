/**
 * plugins/vuetify.ts
 *
 * Framework documentation: https://vuetifyjs.com`
 */

// Styles
import 'material-design-icons-iconfont/dist/material-design-icons.css'
import 'vuetify/styles'
import { aliases, md } from 'vuetify/iconsets/md'

// Composables
import { createVuetify } from 'vuetify'

// https://vuetifyjs.com/en/introduction/why-vuetify/#feature-guides
export default createVuetify({
  theme: {
    defaultTheme: 'dark',
    themes: {
      dark: {
        dark: true,
        colors: {
          primary: '#1eb6c3',
          secondary: '#424242',
          accent: '#82B1FF',
          error: '#FF5252',
          info: '#2196F3',
          success: '#4CAF50',
          warning: '#FFC107'
        }
      },
      light: {
        dark: false,
        colors: {
          primary: '#9B0505',
          secondary: '#424242',
          accent: '#82B1FF',
          error: '#FF5252',
          info: '#2196F3',
          success: '#4CAF50',
          warning: '#FFC107'
        }
      }
    }
  },
  icons: {
    defaultSet: 'md',
    aliases,
    sets: {
      md
    }
  }
})
