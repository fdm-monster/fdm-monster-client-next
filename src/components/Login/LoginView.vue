<template>
  <v-row class="position-relative" no-gutters>
    <v-col cols="12" lg="12">
      <v-container>
        <div class="d-flex align-center" style="min-height: calc(100vh - 148px)">
          <v-row justify="center" class="align-content-center">
            <div class="pt-6 pl-6 d-flex flex-column align-center">
              <Logo/>

              <v-toolbar-title class="text-uppercase text-red">
                <strong>FDM&nbsp;</strong>
                <strong>Monster</strong>
              </v-toolbar-title>

              <v-toolbar-title class="mt-lg-6 mt-sm-5 mb-md-5"> Login to your account</v-toolbar-title>

              <v-card-subtitle class="text-grey mb-md-5">
                Welcome back! Please enter your details
              </v-card-subtitle>
            </div>

            <v-col cols="12" md="12">
              <LoginForm class="login-card"/>
            </v-col>
          </v-row>
        </div>
      </v-container>
    </v-col>
  </v-row>
</template>

<script lang="ts" setup>
import LoginForm from "./LoginForm.vue"
</script>

<style lang="scss">
.login-card {
  max-width: 475px;
  margin: 0 auto;
}

.blur-logo {
  position: absolute;
  filter: blur(18px);
  bottom: 0;
  transform: inherit;
}
</style>
