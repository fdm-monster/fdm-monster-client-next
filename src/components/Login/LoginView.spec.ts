import { mount } from '@vue/test-utils'
import { describe, it, expect, beforeEach } from 'vitest'
import LoginView from '@/components/Login/LoginView.vue' // Adjust the path as necessary
import LoginForm from '@/components/Login/LoginForm.vue'
import Logo from '@/components/Login/Logo.vue'
import { createPinia } from 'pinia'
import { createVuetify } from 'vuetify'
import { createRouter, createWebHistory } from 'vue-router'

const pinia = createPinia()
const vuetify = createVuetify()
// Create a router for testing
const routes = [{ path: '/', component: LoginView }]
const router = createRouter({
  history: createWebHistory(),
  routes
})

describe('LoginView.vue', () => {
  let wrapper: ReturnType<typeof mount>

  beforeEach(() => {
    wrapper = mount(LoginView, {
      global: {
        plugins: [vuetify, pinia, router]
      }
    })
  })

  it('renders the logo component', () => {
    const logo = wrapper.findComponent(Logo)
    expect(logo.exists()).toBe(true)
  })

  it('renders the LoginForm component', () => {
    const loginForm = wrapper.findComponent(LoginForm)
    expect(loginForm.exists()).toBe(true)
  })

  it('displays the correct title', () => {
    const title = wrapper.find('.text-uppercase.text-red') // Using the class selector
    expect(title.exists()).toBe(true) // Check if the title exists
    expect(title.text()).toContain('FDM')
    expect(title.text()).toContain('Monster')
  })

  it('displays the correct subtitle', () => {
    const subtitle = wrapper.find('.v-card-subtitle').text()
    expect(subtitle).toBe('Welcome back! Please enter your details')
  })

  it('renders with the correct structure', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  it('has the correct CSS class for the login card', () => {
    const loginCard = wrapper.find('.login-card')
    expect(loginCard.exists()).toBe(true)
    expect(loginCard.classes()).toContain('login-card')
    expect(loginCard.attributes('style')).toContain('border-radius: 10px')
  })
})
