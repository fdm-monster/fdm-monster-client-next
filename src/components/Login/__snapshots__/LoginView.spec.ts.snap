// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginView.vue > renders with the correct structure 1`] = `
<div
  class="v-row v-row--no-gutters position-relative"
>
  <div
    class="v-col-lg-12 v-col-12"
  >
    <div
      class="v-container v-locale--is-ltr"
    >
      <div
        class="d-flex align-center"
        style="min-height: calc(100vh - 148px);"
      >
        <div
          class="v-row justify-center align-content-center"
        >
          <div
            class="pt-6 pl-6 d-flex flex-column align-center"
          >
            <div
              class="v-responsive v-img v-img--booting shrink mr-1 pt-3 ml-1"
              contain=""
              style="width: 150px; opacity: 0.85;"
            >
              <div
                class="v-responsive__sizer"
              />
              
              
              <transition-stub
                appear="true"
                css="true"
                name="fade-transition"
                persisted="false"
              >
                <img
                  class="v-img__img v-img__img--contain"
                  src="/src/assets/logo.png"
                  style="display: none;"
                />
              </transition-stub>
              <transition-stub
                appear="false"
                css="true"
                name="fade-transition"
                persisted="false"
              >
                <!---->
              </transition-stub>
              <!---->
              <!---->
              <!---->
              
              
              <!---->
            </div>
            <div
              class="v-toolbar-title text-uppercase text-red"
            >
              <div
                class="v-toolbar-title__placeholder"
              >
                <!---->
                
                <strong>
                   FDMÂ  
                </strong>
                <strong>
                   Monster 
                </strong>
                
              </div>
            </div>
            <div
              class="v-toolbar-title mt-lg-6 mt-sm-5 mb-md-5"
            >
              <div
                class="v-toolbar-title__placeholder"
              >
                <!---->
                
                 Login to your account 
                
              </div>
            </div>
            <div
              class="v-card-subtitle text-grey mb-md-5"
            >
               Welcome back! Please enter your details 
            </div>
          </div>
          <div
            class="v-col-md-12 v-col-12"
          >
            <div
              class="v-card v-theme--light v-card--density-default elevation-10 v-card--variant-elevated pa-4 login-card"
              style="border-radius: 10px;"
            >
              <!---->
              <div
                class="v-card__loader"
              >
                <div
                  aria-hidden="true"
                  aria-valuemax="100"
                  aria-valuemin="0"
                  class="v-progress-linear v-theme--light v-locale--is-ltr"
                  role="progressbar"
                  style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                >
                  <!---->
                  <div
                    class="v-progress-linear__background"
                  />
                  <div
                    class="v-progress-linear__buffer"
                    style="width: 0%;"
                  />
                  <transition-stub
                    appear="false"
                    css="true"
                    name="fade-transition"
                    persisted="false"
                  >
                    <div
                      class="v-progress-linear__indeterminate"
                    >
                      
                      <div
                        class="v-progress-linear__indeterminate long"
                      />
                      <div
                        class="v-progress-linear__indeterminate short"
                      />
                      
                    </div>
                  </transition-stub>
                  <!---->
                </div>
              </div>
              <!---->
              <!---->
              
              <div
                class="v-card-text"
              >
                <form
                  class="v-form"
                  novalidate=""
                >
                  
                  <label>
                     Username 
                  </label>
                  <div
                    class="v-input v-input--horizontal v-input--density-default v-theme--light v-locale--is-ltr v-text-field v-input--plain-underlined"
                  >
                    <div
                      class="v-input__prepend"
                    >
                      <!---->
                      <i
                        aria-hidden="true"
                        class="person mdi v-icon notranslate v-theme--light v-icon--size-default"
                      />
                    </div>
                    <div
                      class="v-input__control"
                    >
                      
                      <div
                        class="v-field v-field--variant-underlined v-theme--light v-locale--is-ltr"
                      >
                        <div
                          class="v-field__overlay"
                        />
                        <div
                          class="v-field__loader"
                        >
                          <div
                            aria-hidden="true"
                            aria-valuemax="100"
                            aria-valuemin="0"
                            class="v-progress-linear v-theme--light v-locale--is-ltr"
                            role="progressbar"
                            style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                          >
                            <!---->
                            <div
                              class="v-progress-linear__background"
                            />
                            <div
                              class="v-progress-linear__buffer"
                              style="width: 0%;"
                            />
                            <transition-stub
                              appear="false"
                              css="true"
                              name="fade-transition"
                              persisted="false"
                            >
                              <div
                                class="v-progress-linear__indeterminate"
                              >
                                
                                <div
                                  class="v-progress-linear__indeterminate long"
                                />
                                <div
                                  class="v-progress-linear__indeterminate short"
                                />
                                
                              </div>
                            </transition-stub>
                            <!---->
                          </div>
                        </div>
                        <!---->
                        <div
                          class="v-field__field"
                          data-no-activator=""
                        >
                          <!---->
                          <label
                            class="v-label v-field-label"
                            for="input-0"
                          >
                            <!---->
                            
                            Username
                            
                          </label>
                          
                          
                          <!---->
                          <input
                            aria-describedby="input-0-messages"
                            autofocus=""
                            class="v-field__input"
                            id="input-0"
                            name="login"
                            size="1"
                            type="text"
                            value=""
                          />
                          <!---->
                          
                          
                        </div>
                        <!---->
                        <!---->
                        <div
                          class="v-field__outline"
                        >
                          <!---->
                          <label
                            aria-hidden="true"
                            class="v-label v-field-label v-field-label--floating"
                            for="input-0"
                          >
                            <!---->
                            
                            Username
                            
                          </label>
                        </div>
                      </div>
                      
                    </div>
                    <!---->
                    <div
                      class="v-input__details"
                    >
                      <transition-group-stub
                        appear="false"
                        aria-live="polite"
                        class="v-messages"
                        css="true"
                        id="input-0-messages"
                        name="slide-y-transition"
                        persisted="false"
                        role="alert"
                        tag="div"
                      >
                        <!---->
                      </transition-group-stub>
                      <!---->
                    </div>
                  </div>
                  <div
                    class="v-input v-input--horizontal v-input--density-default v-theme--light v-locale--is-ltr v-text-field v-input--plain-underlined"
                  >
                    <div
                      class="v-input__prepend"
                    >
                      <!---->
                      <i
                        aria-hidden="true"
                        class="lock mdi v-icon notranslate v-theme--light v-icon--size-default"
                      />
                    </div>
                    <div
                      class="v-input__control"
                    >
                      
                      <div
                        class="v-field v-field--appended v-field--variant-underlined v-theme--light v-locale--is-ltr"
                      >
                        <div
                          class="v-field__overlay"
                        />
                        <div
                          class="v-field__loader"
                        >
                          <div
                            aria-hidden="true"
                            aria-valuemax="100"
                            aria-valuemin="0"
                            class="v-progress-linear v-theme--light v-locale--is-ltr"
                            role="progressbar"
                            style="top: 0px; height: 0px; --v-progress-linear-height: 2px;"
                          >
                            <!---->
                            <div
                              class="v-progress-linear__background"
                            />
                            <div
                              class="v-progress-linear__buffer"
                              style="width: 0%;"
                            />
                            <transition-stub
                              appear="false"
                              css="true"
                              name="fade-transition"
                              persisted="false"
                            >
                              <div
                                class="v-progress-linear__indeterminate"
                              >
                                
                                <div
                                  class="v-progress-linear__indeterminate long"
                                />
                                <div
                                  class="v-progress-linear__indeterminate short"
                                />
                                
                              </div>
                            </transition-stub>
                            <!---->
                          </div>
                        </div>
                        <!---->
                        <div
                          class="v-field__field"
                          data-no-activator=""
                        >
                          <!---->
                          <label
                            class="v-label v-field-label"
                            for="input-2"
                          >
                            <!---->
                            
                            Password
                            
                          </label>
                          
                          
                          <!---->
                          <input
                            aria-describedby="input-2-messages"
                            class="v-field__input"
                            id="input-2"
                            name="password"
                            password=""
                            size="1"
                            type="password"
                            value=""
                          />
                          <!---->
                          
                          
                        </div>
                        <!---->
                        <div
                          class="v-field__append-inner"
                        >
                          <!---->
                          <i
                            aria-hidden="false"
                            aria-label="Password appended action"
                            class="visibility mdi v-icon notranslate v-theme--light v-icon--size-default v-icon--clickable"
                            role="button"
                            tabindex="0"
                          />
                        </div>
                        <div
                          class="v-field__outline"
                        >
                          <!---->
                          <label
                            aria-hidden="true"
                            class="v-label v-field-label v-field-label--floating"
                            for="input-2"
                          >
                            <!---->
                            
                            Password
                            
                          </label>
                        </div>
                      </div>
                      
                    </div>
                    <!---->
                    <div
                      class="v-input__details"
                    >
                      <transition-group-stub
                        appear="false"
                        aria-live="polite"
                        class="v-messages"
                        css="true"
                        id="input-2-messages"
                        name="slide-y-transition"
                        persisted="false"
                        role="alert"
                        tag="div"
                      >
                        <!---->
                      </transition-group-stub>
                      <!---->
                    </div>
                  </div>
                  <!--v-if-->
                  <!--v-if-->
                  
                </form>
              </div>
              <div
                class="v-card-actions"
              >
                
                <button
                  class="v-btn v-btn--slim v-theme--light bg-primary v-btn--density-default v-btn--variant-flat pa-4"
                  style="width: 100%;"
                  type="button"
                >
                  
                  <span
                    class="v-btn__overlay"
                  />
                  <span
                    class="v-btn__underlay"
                  />
                  
                  <!---->
                  <span
                    class="v-btn__content"
                    data-no-activator=""
                  >
                    
                    
                     Login 
                    
                    
                  </span>
                  <!---->
                  <!---->
                </button>
                
              </div>
              <div
                class="v-card-actions"
              >
                
                <button
                  class="v-btn v-btn--disabled v-btn--slim v-theme--light v-btn--density-default v-btn--variant-flat pa-4"
                  disabled=""
                  style="width: 100%;"
                  type="button"
                >
                  
                  <span
                    class="v-btn__overlay"
                  />
                  <span
                    class="v-btn__underlay"
                  />
                  
                  <!---->
                  <span
                    class="v-btn__content"
                    data-no-activator=""
                  >
                    
                    
                     Register new account (not enabled) 
                    <i
                      aria-hidden="true"
                      class="right mdi v-icon notranslate v-theme--light v-icon--size-default pl-5"
                    />
                    
                    
                  </span>
                  <!---->
                  <!---->
                </button>
                
              </div>
              
              <!---->
              
              <!---->
              <span
                class="v-card__underlay"
              />
              
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
